pkg/check/table_structure.go

```diff
@@ -77,73 +156,283 @@ func (c *TablesChecker) Name() string {
-			return errors.Trace(err)
+		option := c.checkTableOption(opt)
+		if option != nil {
+			options = append(options, option)
 		}
 	}
-	return nil
+	return options
 }
 
-func (c *TablesChecker) checkColumnDef(def *ast.ColumnDef) error {
+func (c *TablesChecker) checkColumnDef(def *ast.ColumnDef) *incompatibilityOption {
 	return nil
 }
 
-func (c *TablesChecker) checkConstraint(cst *ast.Constraint) error {
+func (c *TablesChecker) checkConstraint(cst *ast.Constraint) *incompatibilityOption {
 	switch cst.Tp {
 	case ast.ConstraintForeignKey:
-		c.warnings++
-		return errors.Errorf("Foreign Key is parsed but ignored by TiDB.")
+		return &incompatibilityOption{
+			state:       StateWarning,
+			instruction: fmt.Sprintf("please ref document: https://github.com/pingcap/docs-cn/blob/master/sql/ddl.md"),
+			errMessage:  fmt.Sprintf("Foreign Key %s is parsed but ignored by TiDB.", cst.Name),
+		}
 	}
+
 	return nil
 }
 
-func (c *TablesChecker) checkTableOption(opt *ast.TableOption) error {
+func (c *TablesChecker) checkTableOption(opt *ast.TableOption) *incompatibilityOption {
 	switch opt.Tp {
 	case ast.TableOptionCharset:
 		// Check charset
 		cs := strings.ToLower(opt.StrValue)
 		if cs != "binary" && !charset.ValidCharsetAndCollation(cs, "") {
-			c.errors++
-			return errors.Errorf("Unsupported charset %s", opt.StrValue)
+			return &incompatibilityOption{
+				state:       StateFailure,
+				instruction: fmt.Sprintf("please ref document: https://github.com/pingcap/docs-cn/blob/master/sql/character-set-support.md"),
+				errMessage:  fmt.Sprintf("unsupport charset %s", opt.StrValue),
+			}
+		}
+	}
+	return nil
+}
+
+// ShardingTablesCheck checks consistency of table structures of one sharding group
+// * check whether they have same column list
+// * check whether they have auto_increment key
+type ShardingTablesCheck struct {
+	name string
+
+	dbs    map[string]*sql.DB
+	tables map[string]map[string][]string // instance => {schema: [table1, table2, ...]}
+}
+
+// NewShardingTablesCheck returns a Checker
+func NewShardingTablesCheck(name string, dbs map[string]*sql.DB, tables map[string]map[string][]string) Checker {
+	return &ShardingTablesCheck{
+		name:   name,
+		dbs:    dbs,
+		tables: tables,
+	}
+}
+
+// Check implements Checker interface
+func (c *ShardingTablesCheck) Check(ctx context.Context) *Result {
+	r := &Result{
+		Name:  c.Name(),
+		Desc:  "check consistency of sharding table structures",
+		State: StateSuccess,
+		Extra: fmt.Sprintf("sharding %s", c.name),
+	}
+
+	var (
+		hasAutoIncrementKeyTableNum int
+		hasAutoIncrementKeyTables   = make(map[string][]string)
+
+		stmtNode  *ast.CreateTableStmt
+		tableName string
+	)
+	for instance, schemas := range c.tables {
+		db, ok := c.dbs[instance]
+		if !ok {
+			markCheckError(r, errors.NotFoundf("client for instance %s", instance))
+			return r
+		}
+
+		for schema, tables := range schemas {
+			for _, table := range tables {
+				if len(tables) == 0 {
+					continue
```

Comment:
if `len(tables) == 0`, then it will NOT go into this `range` block?

Reply:
obsolete code